src/Desktop.Analyzers/UnitTests/DoNotUseInsecureDTDProcessingUseXmlReaderForSchemaReadTests.cs

```diff
@@ -0,0 +1,410 @@
+        End Try
+    End Sub
+End Class",
+                GetCA3075BasicResultAt(13, 39, "TestMethod")
+            );
+        }
+
+        [Fact]
+        public void UseXmlSchemaReadInAsyncAwaitShouldGenerateDiagnostic()
+        {
+            VerifyCSharp(@"
+ using System.Threading.Tasks;
+using System.Data;
+using System.IO;
+using System.Xml.Schema;
+    class TestClass
+    {
+        private async Task TestMethod()
+        {
+            await Task.Run(() => {
+                var src = """";
+                TextReader tr = new StreamReader(src);
+                XmlSchema schema = XmlSchema.Read(tr, null);
+            });
+        }
+
+        private async void TestMethod2()
+        {
+            await TestMethod();
+        }
+    }",
+                GetCA3075CSharpResultAt(13, 36, "Run")
+            );
+
+            VerifyBasic(@"
+Imports System.Threading.Tasks
+Imports System.Data
+Imports System.IO
+Imports System.Xml.Schema
+Class TestClass
+    Private Function TestMethod() As Task
+        Await Task.Run(Function() 
+        Dim src = """"
+        Dim tr As TextReader = New StreamReader(src)
+        Dim schema As XmlSchema = XmlSchema.Read(tr, Nothing)
+
+End Function)
+    End Function
+
+    Private Sub TestMethod2()
+        Await TestMethod()
+    End Sub
+End Class",
+                GetCA3075BasicResultAt(11, 35, "Run")
+            );
+        }
+
+        [Fact]
+        public void UseXmlSchemaReadInDelegateShouldGenerateDiagnostic()
+        {
+            VerifyCSharp(@"
+using System.Data;
+using System.IO;
+using System.Xml.Schema;
+    class TestClass
+    {
+        delegate void Del();
+
+        Del d = delegate () {
+            var src = """";
+            TextReader tr = new StreamReader(src);
+            XmlSchema schema = XmlSchema.Read(tr, null);
+        };
+    }",
+                GetCA3075CSharpResultAt(12, 32, "TestClass")
+            );
+
+            VerifyBasic(@"
+Imports System.Data
+Imports System.IO
+Imports System.Xml.Schema
+Class TestClass
+    Private Delegate Sub Del()
+
+    Private d As Del = Sub() 
+    Dim src = """"
+    Dim tr As TextReader = New StreamReader(src)
+    Dim schema As XmlSchema = XmlSchema.Read(tr, Nothing)
+
+End Sub
+End Class",
+                GetCA3075BasicResultAt(11, 31, "TestClass")
+            );
+        }
+
+        [Fact]
+        public void UseXmlSchemaReadWithXmlReaderShouldNotGenerateDiagnostic()
+        {
+            VerifyCSharp(@"
+using System.Xml;
+using System.Xml.Schema;
+
+namespace FxCopUnsafeXml
+{
+    public class UseXmlReaderForSchemaRead
+    {
+        public void TestMethod19(XmlTextReader reader)
+        {
+            XmlSchema schema = XmlSchema.Read(reader, null);
+        }
+    }
+}"
+            );
+
+            VerifyBasic(@"
+Imports System.Xml
+Imports System.Xml.Schema
+
+Namespace FxCopUnsafeXml
+    Public Class UseXmlReaderForSchemaRead
+        Public Sub TestMethod19(reader As XmlTextReader)
```

Comment:
You're not really safe unless `XmlTextReader.DtdProcessing` is set to `Prohibit`, but I guess you can't be sure of that without doing a data flow analysis, so I guess the best we can do is to ensure that _some_ `XmlTextReader` object is being used, right?